### ISSUE TITLE
Kuby Ingress Error: Could not apply https://raw.githubusercontent.com/kubernetes/ingress-nginx....

### ISSUE THREAD

**Issue by scart88** [getkuby/kuby-core#94]
I'm getting this error when we are trying to setup the provider on Digital Ocean.

## Kuby config
```ruby
require 'active_support/core_ext'
require 'active_support/encrypted_configuration'

# Define a production Kuby deploy environment
Kuby.define('KubyApp') do
  environment(:production) do
    # Because the Rails environment isn't always loaded when
    # your Kuby config is loaded, provide access to Rails
    # credentials manually.
    app_creds = ActiveSupport::EncryptedConfiguration.new(
      config_path: File.join('config', 'credentials', 'production.yml.enc'),
      key_path: File.join('config', 'credentials', 'production.key'),
      env_key: 'RAILS_MASTER_KEY',
      raise_if_missing_key: true
    )

    docker do
      # Configure your Docker registry credentials here. Add them to your
      # Rails credentials file by running `bundle exec rake credentials:edit`.
      credentials do
        username app_creds[:KUBY_DOCKER_USERNAME]
        password app_creds[:KUBY_DOCKER_PASSWORD]
        email app_creds[:KUBY_DOCKER_EMAIL]
      end

      # distro :alpine

      # Configure the URL to your Docker image here, eg:
      # image_url 'foo.bar.com/me/myproject'
      #
      # If you're using Gitlab's Docker registry, try something like this:
      image_url 'registry.gitlab.com/user/repo'
    end

    kubernetes do

      provider :digitalocean do
        access_token app_creds[:DIGITALOCEAN_ACCESS_TOKEN]
        cluster_id app_creds[:CLUSTER_ID]
      end

      # Add a plugin that facilitates deploying a Rails app.
      add_plugin :rails_app do
        hostname 'app2.domain.com'
        manage_database false

        env do
          data do
            add "DATABASE_URL", app_creds[:DATABASE_URL]
          end
        end
        # configure database credentials
        # database do
        #   user app_creds[:KUBY_DB_USER]
        #   password app_creds[:KUBY_DB_PASSWORD]
        # end
      end

      # Use Docker Desktop as the provider.
      # See: https://www.docker.com/products/docker-desktop
      #
      # Note: you will likely want to use a different provider when deploying
      # your application into a production environment. To configure a different
      # provider, add the corresponding gem to your gemfile and update the
      # following line according to the provider gem's README.

     
    end
  end
end

```

## Kuby terminal error on kuby setup
```
bundle exec kuby -e production setup
Refreshing kubeconfig...
Successfully refreshed kubeconfig!
Deploying nginx ingress resources
Error from server (NotFound): namespaces "ingress-nginx" not found
namespace/ingress-nginx created
configmap/nginx-configuration created
configmap/tcp-services created
configmap/udp-services created
serviceaccount/nginx-ingress-serviceaccount created
deployment.apps/nginx-ingress-controller created
limitrange/ingress-nginx created
unable to recognize "https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml": no matches for kind "Role" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
Could not apply https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml: kubectl exited with status code 1
error: Could not apply https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml: kubectl exited with status code 1

```

## Kuby terminal error on kuby deploy
```
bundle exec kuby -e production deploy
Error from server (NotFound): namespaces "app-production" not found
Validating global resource, namespace 'app-production'
namespace/app-production created (dry run)
Deploying namespace 'app-production'
namespace/app-production created
[INFO][2022-03-11 17:41:57 +0200]
[INFO][2022-03-11 17:41:57 +0200]	------------------------------------Phase 1: Initializing deploy------------------------------------
[INFO][2022-03-11 17:42:00 +0200]	All required parameters and files are present
[INFO][2022-03-11 17:42:00 +0200]	Discovering resources:
[INFO][2022-03-11 17:42:03 +0200]	  - Service/app-assets-svc
[INFO][2022-03-11 17:42:03 +0200]	  - Ingress/app-ingress
[INFO][2022-03-11 17:42:03 +0200]	  - Deployment/app-assets
[INFO][2022-03-11 17:42:03 +0200]	  - ConfigMap/app-config
[INFO][2022-03-11 17:42:03 +0200]	  - ConfigMap/app-assets-nginx-config
[INFO][2022-03-11 17:42:03 +0200]	  - ServiceAccount/app-assets-sa
[INFO][2022-03-11 17:42:03 +0200]	  - Service/app-svc
[INFO][2022-03-11 17:42:03 +0200]	  - ClusterIssuer/letsencrypt-production
[INFO][2022-03-11 17:42:03 +0200]	  - ServiceAccount/app-sa
[INFO][2022-03-11 17:42:03 +0200]	  - Deployment/app-web
[INFO][2022-03-11 17:42:03 +0200]	  - Secret/app-secrets
[INFO][2022-03-11 17:42:03 +0200]	  - Secret/app-registry-secret
[INFO][2022-03-11 17:42:07 +0200]
[INFO][2022-03-11 17:42:07 +0200]	------------------------------------------Result: FAILURE-------------------------------------------
[FATAL][2022-03-11 17:42:07 +0200]	Template validation failed
[FATAL][2022-03-11 17:42:07 +0200]
[FATAL][2022-03-11 17:42:07 +0200]	Invalid template: ClusterIssuer-letsencrypt-production20220311-21350-rc7rha.yml
[FATAL][2022-03-11 17:42:07 +0200]	> Error message:
[FATAL][2022-03-11 17:42:07 +0200]	    W0311 17:42:03.089014   21380 helpers.go:557] --dry-run is deprecated and can be replaced with --dry-run=client.
[FATAL][2022-03-11 17:42:07 +0200]	    error: unable to recognize "/var/folders/hh/z_vjqk3j3dl5vw0whd7h7bp80000gn/T/ClusterIssuer-letsencrypt-production20220311-21350-rc7rha.yml": no matches for kind "ClusterIssuer" in version "cert-manager.io/v1alpha2"
[FATAL][2022-03-11 17:42:07 +0200]	> Template content:
[FATAL][2022-03-11 17:42:07 +0200]	    ---
[FATAL][2022-03-11 17:42:07 +0200]	    apiVersion: cert-manager.io/v1alpha2
[FATAL][2022-03-11 17:42:07 +0200]	    kind: ClusterIssuer
[FATAL][2022-03-11 17:42:07 +0200]	    metadata:
[FATAL][2022-03-11 17:42:07 +0200]	      name: letsencrypt-production
[FATAL][2022-03-11 17:42:07 +0200]	      namespace: cert-manager
[FATAL][2022-03-11 17:42:07 +0200]	    spec:
[FATAL][2022-03-11 17:42:07 +0200]	      acme:
[FATAL][2022-03-11 17:42:07 +0200]	        server: https://acme-v02.api.letsencrypt.org/directory
[FATAL][2022-03-11 17:42:07 +0200]	        email: shopandfy@xmail8.com
[FATAL][2022-03-11 17:42:07 +0200]	        privateKeySecretRef:
[FATAL][2022-03-11 17:42:07 +0200]	          name: letsencrypt-production
[FATAL][2022-03-11 17:42:07 +0200]	        solvers:
[FATAL][2022-03-11 17:42:07 +0200]	        - http01:
[FATAL][2022-03-11 17:42:07 +0200]	            ingress:
[FATAL][2022-03-11 17:42:07 +0200]	              class: nginx
[FATAL][2022-03-11 17:42:07 +0200]
[FATAL][2022-03-11 17:42:07 +0200]
[FATAL][2022-03-11 17:42:07 +0200]	Invalid template: Ingress-app-ingress20220311-21350-ht7nke.yml
[FATAL][2022-03-11 17:42:07 +0200]	> Error message:
[FATAL][2022-03-11 17:42:07 +0200]	    W0311 17:42:05.241076   21398 helpers.go:557] --dry-run is deprecated and can be replaced with --dry-run=client.
[FATAL][2022-03-11 17:42:07 +0200]	    error: unable to recognize "/var/folders/hh/z_vjqk3j3dl5vw0whd7h7bp80000gn/T/Ingress-app-ingress20220311-21350-ht7nke.yml": no matches for kind "Ingress" in version "extensions/v1beta1"
[FATAL][2022-03-11 17:42:07 +0200]	> Template content:
[FATAL][2022-03-11 17:42:07 +0200]	    ---
[FATAL][2022-03-11 17:42:07 +0200]	    apiVersion: extensions/v1beta1
[FATAL][2022-03-11 17:42:07 +0200]	    kind: Ingress
[FATAL][2022-03-11 17:42:07 +0200]	    metadata:
[FATAL][2022-03-11 17:42:07 +0200]	      annotations:
[FATAL][2022-03-11 17:42:07 +0200]	        kubernetes.io/ingress.class: nginx
[FATAL][2022-03-11 17:42:07 +0200]	        cert-manager.io/cluster-issuer: letsencrypt-production
[FATAL][2022-03-11 17:42:07 +0200]	      name: app-ingress
[FATAL][2022-03-11 17:42:07 +0200]	      namespace: app-production
[FATAL][2022-03-11 17:42:07 +0200]	    spec:
[FATAL][2022-03-11 17:42:07 +0200]	      rules:
[FATAL][2022-03-11 17:42:07 +0200]	      - host: app2.domain.com
[FATAL][2022-03-11 17:42:07 +0200]	        http:
[FATAL][2022-03-11 17:42:07 +0200]	          paths:
[FATAL][2022-03-11 17:42:07 +0200]	          - backend:
[FATAL][2022-03-11 17:42:07 +0200]	              serviceName: app-svc
[FATAL][2022-03-11 17:42:07 +0200]	              servicePort: 8080
[FATAL][2022-03-11 17:42:07 +0200]	            path: "/"
[FATAL][2022-03-11 17:42:07 +0200]	      - host: app2.domain.com
[FATAL][2022-03-11 17:42:07 +0200]	        http:
[FATAL][2022-03-11 17:42:07 +0200]	          paths:
[FATAL][2022-03-11 17:42:07 +0200]	          - backend:
[FATAL][2022-03-11 17:42:07 +0200]	              serviceName: app-assets-svc
[FATAL][2022-03-11 17:42:07 +0200]	              servicePort: 8082
[FATAL][2022-03-11 17:42:07 +0200]	            path: "/assets"
[FATAL][2022-03-11 17:42:07 +0200]	          - backend:
[FATAL][2022-03-11 17:42:07 +0200]	              serviceName: app-assets-svc
[FATAL][2022-03-11 17:42:07 +0200]	              servicePort: 8082
[FATAL][2022-03-11 17:42:07 +0200]	            path: "/packs"
[FATAL][2022-03-11 17:42:07 +0200]	      tls:
[FATAL][2022-03-11 17:42:07 +0200]	      - hosts:
[FATAL][2022-03-11 17:42:07 +0200]	        - app2.domain.com
[FATAL][2022-03-11 17:42:07 +0200]	        secretName: app-tls
[FATAL][2022-03-11 17:42:07 +0200]
error: Template validation failed
```

Is there anything I can do to fix it?

**Comment by scart88** [comment]
Thank you very much for your fast reply and help.
I've updated the gem master version. 

I'm unable to select version v.1.19, is not showing in the DigitalOcean dropdown anymore.
I can only select v.1.20, v1.21 and v1.22.

Right now, my Kubernetes version is v1.22.7. 
However, with this gem version, I'm getting the following errors
```
> bundle exec kuby -e production build
error: undefined method `map' for nil:NilClass

> bundle exec kuby -e production push
error: undefined method `map' for nil:NilClass
```

Setup seems to work better than before
```
bundle exec kuby -e production setup
Refreshing kubeconfig...
Successfully refreshed kubeconfig!
Deploying nginx ingress resources
Error from server (NotFound): namespaces "ingress-nginx" not found
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
configmap/ingress-nginx-controller created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
service/ingress-nginx-controller-admission created
service/ingress-nginx-controller created
deployment.apps/ingress-nginx-controller created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
serviceaccount/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
Nginx ingress resources deployed!
Installing cert-manager...
namespace/cert-manager created
serviceaccount/cert-manager-cainjector created
serviceaccount/cert-manager created
serviceaccount/cert-manager-webhook created
clusterrole.rbac.authorization.k8s.io/cert-manager-webhook:webhook-requester created
clusterrole.rbac.authorization.k8s.io/cert-manager-view created
clusterrole.rbac.authorization.k8s.io/cert-manager-edit created
service/cert-manager created
service/cert-manager-webhook created
deployment.apps/cert-manager-cainjector created
deployment.apps/cert-manager created
deployment.apps/cert-manager-webhook created
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "Role" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "Role" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "MutatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1"
unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ValidatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1"
Could not apply https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml: kubectl exited with status code 1
error: Could not apply https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml: kubectl exited with status code 1
```

I will try with v.1.20. Let's see if it works.

**Comment by denikus** [comment]
> Thank you very much for your fast reply and help. I've updated the gem master version.
> 
> I'm unable to select version v.1.19, is not showing in the DigitalOcean dropdown anymore. I can only select v.1.20, v1.21 and v1.22.
> 
> Right now, my Kubernetes version is v1.22.7. However, with this gem version, I'm getting the following errors
> 
> ```
> > bundle exec kuby -e production build
> error: undefined method `map' for nil:NilClass
> 
> > bundle exec kuby -e production push
> error: undefined method `map' for nil:NilClass
> ```
> 
> Setup seems to work better than before
> 
> ```
> bundle exec kuby -e production setup
> Refreshing kubeconfig...
> Successfully refreshed kubeconfig!
> Deploying nginx ingress resources
> Error from server (NotFound): namespaces "ingress-nginx" not found
> namespace/ingress-nginx created
> serviceaccount/ingress-nginx created
> configmap/ingress-nginx-controller created
> clusterrole.rbac.authorization.k8s.io/ingress-nginx created
> clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
> role.rbac.authorization.k8s.io/ingress-nginx created
> rolebinding.rbac.authorization.k8s.io/ingress-nginx created
> service/ingress-nginx-controller-admission created
> service/ingress-nginx-controller created
> deployment.apps/ingress-nginx-controller created
> ingressclass.networking.k8s.io/nginx created
> validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
> serviceaccount/ingress-nginx-admission created
> clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
> clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
> role.rbac.authorization.k8s.io/ingress-nginx-admission created
> rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
> job.batch/ingress-nginx-admission-create created
> job.batch/ingress-nginx-admission-patch created
> Nginx ingress resources deployed!
> Installing cert-manager...
> namespace/cert-manager created
> serviceaccount/cert-manager-cainjector created
> serviceaccount/cert-manager created
> serviceaccount/cert-manager-webhook created
> clusterrole.rbac.authorization.k8s.io/cert-manager-webhook:webhook-requester created
> clusterrole.rbac.authorization.k8s.io/cert-manager-view created
> clusterrole.rbac.authorization.k8s.io/cert-manager-edit created
> service/cert-manager created
> service/cert-manager-webhook created
> deployment.apps/cert-manager-cainjector created
> deployment.apps/cert-manager created
> deployment.apps/cert-manager-webhook created
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "Role" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "Role" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "RoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRole" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ClusterRoleBinding" in version "rbac.authorization.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "MutatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1"
> unable to recognize "https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml": no matches for kind "ValidatingWebhookConfiguration" in version "admissionregistration.k8s.io/v1beta1"
> Could not apply https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml: kubectl exited with status code 1
> error: Could not apply https://github.com/jetstack/cert-manager/releases/download/v0.13.1/cert-manager.yaml: kubectl exited with status code 1
> ```
> 
> I will try with v.1.20. Let's see if it works.

Did you manage to make it work? Exactly same problem here. Stuck on cert-manager now.

**Comment by denikus** [comment]
I've downgraded Kubernetes cluster to version 1.20 and it helps.

**Comment by scart88** [comment]
Thanks. Setup `bundle exec kuby -e production setup` command works on v1.20.
However, the deploy command is getting stuck at `-create-db` on my end.

**Comment by denikus** [comment]
> Thanks. Setup `bundle exec kuby -e production setup` command works on v1.20. However, the deploy command is getting stuck at `-create-db` on my end.

Exactly same problem, fighting it right now. Let me know if you solve it, please.

**Comment by scart88** [comment]
Not yet, but I found this relevant issue #74. I'm kind of stuck, so I took a pause on trying k8s. 
Please let me know if you find a way to make it work.

**Comment by scart88** [comment]
Unfortunately it didn't worked. 

```
NAME                                           READY   STATUS                  RESTARTS   AGE
cm-acme-http-solver-stpkj                      1/1     Running                 0          50s
hotwiresample-assets-77cbcb497c-kq6pg   1/1     Running                 0          4m1s
hotwiresample-web-74df7f7d9b-ngkgh      0/1     Init:CrashLoopBackOff   3          4m1s
❯ bundle exec kuby -e production kubectl -N -- logs -c app-create-db hotwiresample-web-74df7f7d9b-ngkgh

error: container app-create-db is not valid for pod hotwiresample-web-74df7f7d9b-ngkgh
```

**Comment by camertron** [comment]
Hey @scart88, glad you were able to get things working!

> It took some time, and finally the SSL certificate did his thing and now it works.

Yeah, sometimes certs are slow to issue because of DNS propagation lag. Unfortunately I don't know of any workarounds for that, but 1) usually it's just a matter of waiting for a bit, and 2) you should only have to worry about it once.

> Considering it took so much time to figure out how to make this to work. I'm a little afraid to go in production with this. I don't want to abuse the issues and your time every time I get an error.

Yes, I totally understand. Kuby is pre-1.0 software, and I'm still actively working out the kinks. I really appreciate you taking the time to give it a try and report your experience back to me. It's super helpful to see where folks are getting tripped up.

> Do you have any recommendations on what would be my next steps in learning more?

That's a good question. What specifically are you interested in learning more about? Kuby uses Docker and Kubernetes under the hood. I'd be happy to point you at some resources for either or both of those things. A good place to start for now might be to run `kuby dockerfiles` and `kuby resources`. Those will print out the Dockerfiles and Kubernetes resource manifests Kuby will use to build and deploy your app. They might not make sense right away, but it's good to know what parts of Docker and k8s Kuby is using so you can narrow your focus.

> Do I have to worry about rollbacks, log rotation, backups (I already have backups on the DO managed db), cron jobs?

Rollbacks are baked into Kuby. You should be able to run `kuby rollback` to deploy the previous release.

In Kubernetes, logs are automatically sent to STDOUT. I believe k8s will keep a certain number of lines around if you want to see log output later, but other than that logs are simply discarded. So no log rotation is necessary as far as I know. If you're interested in looking at old logs, I would recommend a log aggregation service like SumoLogic, etc.

You don't have to worry about backups unless you're writing to the local disk for some reason. If you are, I believe DigitalOcean has a backup feature you could look into.

I'm not sure what you mean about cron jobs? Kubernetes has a cron feature that Kuby doesn't expose.

**Comment by scart88** [comment]
> Yeah, sometimes certs are slow to issue because of DNS propagation lag. Unfortunately I don't know of any workarounds for that, but 1) usually it's just a matter of waiting for a bit, and 2) you should only have to worry about it once.

I don't much about the Kubernetes ingress, but I found it so easy to run the caddy docker file and do a reverse proxy like in the example below. However, I guess is much more complicated in Kubernetes.

```
docker run -d --name p-caddy --restart always --network host -v /home/Caddyfile:/etc/caddy/Caddyfile -v /home/caddy-data:/data caddy
```

 Caddyfile
``` 
example-domain.com {
    reverse_proxy 0.0.0.0:3000
}
```

> Yes, I totally understand. Kuby is pre-1.0 software, and I'm still actively working out the kinks. I really appreciate you taking the time to give it a try and report your experience back to me. It's super helpful to see where folks are getting tripped up.

I will try to help if I can, however my Rails / Ruby experience is pretty basic. I'm really glad you built this and I have the chance to play with it.

> That's a good question. What specifically are you interested in learning more about? Kuby uses Docker and Kubernetes under the hood. I'd be happy to point you at some resources for either or both of those things. A good place to start for now might be to run kuby dockerfiles and kuby resources. Those will print out the Dockerfiles and Kubernetes resource manifests Kuby will use to build and deploy your app. They might not make sense right away, but it's good to know what parts of Docker and k8s Kuby is using so you can narrow your focus.

Thanks. I got a basic docker course from Nick Janetakis, however didn't had much time to go through everything.

> Rollbacks are baked into Kuby. You should be able to run kuby rollback to deploy the previous release.

That handy. I notice on every kuby build, the registry on Gitlab is growing bigger. Should I remove old images?

> I'm not sure what you mean about cron jobs? Kubernetes has a cron feature that Kuby doesn't expose.

I will have to play with the `whenever gem` and see if it works.

**Comment by denikus** [comment]
@camertron maybe you can help. Seems like Puma just crashing for some db reason. No details even in pods' logs.

` Logs from container 'stopdonaterussiaapi-web' (last 25 lines shown):
[FATAL][2022-03-15 11:33:17 +0200]	      /usr/local/bundle/gems/net-protocol-0.1.2/lib/net/protocol.rb:504: warning: previous definition of Socket was here
[FATAL][2022-03-15 11:33:17 +0200]	      [35447] ! Unable to start worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35447] /usr/local/bundle/gems/activerecord-7.0.2.2/lib/active_record/enum.rb:301:in `raise_conflict_error'
[FATAL][2022-03-15 11:33:17 +0200]	      [35447] Early termination of worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35453] ! Unable to start worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35453] /usr/local/bundle/gems/activerecord-7.0.2.2/lib/active_record/enum.rb:301:in `raise_conflict_error'
[FATAL][2022-03-15 11:33:17 +0200]	      [35453] Early termination of worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35450] ! Unable to start worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35450] /usr/local/bundle/gems/activerecord-7.0.2.2/lib/active_record/enum.rb:301:in `raise_conflict_error'
[FATAL][2022-03-15 11:33:17 +0200]	      [35450] Early termination of worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35456] ! Unable to start worker
[FATAL][2022-03-15 11:33:17 +0200]	      [35456] /usr/local/bundle/gems/activerecord-7.0.2.2/lib/active_record/enum.rb:301:in `raise_conflict_error'
[FATAL][2022-03-15 11:33:17 +0200]	      [35456] Early termination of worker
[FATAL][2022-03-15 11:33:17 +0200]	      /usr/local/lib/ruby/2.7.0/net/protocol.rb:66: warning: already initialized constant Net::ProtocRetryError`

What could it possibly be?

Database in separate db cluster and connection open.

**Comment by denikus** [comment]
Solved. Problem was in DATABASE_URL and how I send it to database.yml.

**Comment by scart88** [comment]
> Yeah, sometimes certs are slow to issue because of DNS propagation lag. Unfortunately I don't know of any workarounds for that, but 1) usually it's just a matter of waiting for a bit, and 2) you should only have to worry about it once.

I found that it works faster with `kind: Issuer` (about 1 or 2 minutes), however it must live in the same namespace with the `App / Service / Deployment / Pod`

This is on DO with a Load Balancer and ingress-nginx helm

```
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-nginx
  namespace: app-production-namespace <--- the issuer is is the same namespace
spec:
  # ACME issuer configuration
  # `email` - the email address to be associated with the ACME account (make sure it's a valid one)
  # `server` - the URL used to access the ACME server’s directory endpoint
  # `privateKeySecretRef` - Kubernetes Secret to store the automatically generated ACME account private key
  acme:
    email: email...
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-nginx-private-key
    solvers:
      # Use the HTTP-01 challenge provider
      - http01:
          ingress:
            class: nginx
```

```
apiVersion: v1
kind: Service
metadata:
  name: app-production-service
  namespace: app-production-namespace
spec:
  selector:
    app: app-production
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8080   <--- the ingress is targeting this port
```
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: app-production-ingress
  namespace: app-production-namespace
  annotations:
    cert-manager.io/issuer: letsencrypt-nginx
spec:
  ingressClassName: nginx
  tls:
  - hosts:
    - example.com
    secretName: letsencrypt-nginx
  rules:
  - host: example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: app-production-service
            port:
              number: 8080
```

**Comment by JorgenIvarsson** [comment]
I have the same problem on Linode. Error when running bundle exec kuby -e production setup. I get the "services ingress-nginx not found"-error.  I don't seem to be able to downgrade Kubernetes to a lower version than 1.22 on Linode. As I understand it, there's no workaround for this scenario?

**Comment by camertron** [comment]
Hey @JorgenIvarsson, thanks for your comment. I'm getting ready to publish a big Kuby release ahead of my talk at RailsConf that should address this issue (and a bunch of others). If you can wait a week or so, the upcoming release should unblock you.

**Comment by camertron** [comment]
@JorgenIvarsson yes, it's quite a lot to learn and manage. There are a number of other deployment solutions out there that might work for you - have you considered Heroku for example?

**Comment by scart88** [comment]
@JorgenIvarsson, if Heroku is not an option, you can also check Hatchbox, developed by Chris Oliver from GoRails. It's far easier to get started as a new Rails developer than something like Kubernetes.

**Comment by JorgenIvarsson** [comment]
Thank you for your helpful suggestions @camertron and @scart88 . I am familiar with these services. I have tested Heroku, but not yet Hatchbox.

Heroku seems to be the suggested solution in 90% in all discussions I have read. The other 10% seem to be varying degrees of complicated setup of a VM and then using Capistrano with application servers such as Puma/Passenger/Unicorn together with Postgresql on that VM. So, I must choose between Heroku or a lot of tinkering (If I don’t want to pay Hatchbox to take care of the tinkering). 

But what I want is a solution where I can follow along when DHH builds his famous demo blog, and where I can continue the demo with a deployment that is as easy as it was creating the blog. 

Do you remember when you first saw Rail's scaffolding? Pure magic! 

I want the deployment of an app to give the same feeling of magic (But using industry standard solutions like Docker and Kubernetes) In short, Active Deployment.

Do you think that Kuby could be that solution for a developer of small and simple rails apps (typically a monolithic rails app + redis + postgresql )? Or should I look elsewhere?

**Comment by camertron** [comment]
> But what I want is a solution where I can follow along when DHH builds his famous demo blog, and where I can continue the demo with a deployment that is as easy as it was creating the blog.

Me too!

> Do you think that Kuby could be that solution for a developer of small and simple rails apps (typically a monolithic rails app + redis + postgresql )? Or should I look elsewhere?

That's exactly what I want Kuby to be. It's amazing to me that a large number of Rails developers are still using Capistrano in 2022. Don't get me wrong, Capistrano is a great tool, but I believe there are better ways to deploy apps these days.

**Comment by camertron** [comment]
Hey @JorgenIvarsson, good news! At the end of last week I released the promised big update. Let me know if it works better for you :) Here's the [changelog entry](https://github.com/getkuby/kuby-core/blob/master/CHANGELOG.md#0180) if you're curious.

**Comment by JorgenIvarsson** [comment]
Thanks for the info @camertron! But I am following this project closely and have already upgraded to the new version. I was (and still is) unsure of the best way to  upgrade the gem files and the kuby core. I think that I got it right, but it would be nice to have an instruction on how to upgrade from one version to the next. 
Anyhow, when I continued the instruction I encountered another problem. When I ran _bundle exec kuby -e production setup_ I got ..Refreshing kubeconfig...
And then the error: 404 Not Found: NOT FOUND. 
I have no clue what this means. I know I have been follow along a guide on Kubernetes, using googles gcloud tool and kubectl. Maybe this has altered something in the kubeconf? Anyway I am going to start over with a completely new app and see if I succeed with a fresh project (and credentials). If that doesn't work I will look in to this kubeconfig-file and see if there is something wrong or missing in that file.

**Comment by camertron** [comment]
> I am following this project closely and have already upgraded to the new version.

🎉 🎉 

> I was (and still is) unsure of the best way to upgrade the gem files and the kuby core.

That's good feedback, Kuby does need to have a better upgrade story. I'll have an answer for that before v1.0 (whenever that is).

> When I ran bundle exec kuby -e production setup I got ..Refreshing kubeconfig...
And then the error: 404 Not Found: NOT FOUND.

Hmm, did Kuby print any message saying it had successfully refreshed the kubeconfig? If not, then it sounds like a problem communicating with the cloud provider. Which cloud provider are you using, and are your credentials correct? Kuby doesn't support Google k8s yet. In any case, sounds like an opportunity for a better error message.

**Comment by JorgenIvarsson** [comment]
Thanks for the tip @camertron! Yes,  it turns out when I left off I hadn't fully configured the kubernets section. But once I manage to deploy i got a 404 nginx error message. Tried both with Linode and DigitalOcean. My app did not use a database and maybe that caused some problems? 

Anyhow, I created a new simple app from scratch, this time using a database (just in case using a database is a prerequisite). But when I run "bundle exec kuby -e production build" i get an error: **Kuby does not support the 'postgresql' database adapter.**  This error message really surprised me. I thought that Postgresql was one of the most popular databases in the Rails community? Or is it that there is more than one database adapter for postgres, and I am using the wrong one?

**Comment by camertron** [comment]
> But once I manage to deploy i got a 404 nginx error message. Tried both with Linode and DigitalOcean. My app did not use a database and maybe that caused some problems?

Hmm interesting. How were you attempting to make requests to the app? Was there TLS involved? Enabling/disabling the managed database feature shouldn't be able to cause a connectivity issue like that.

> This error message really surprised me. I thought that Postgresql was one of the most popular databases in the Rails community? Or is it that there is more than one database adapter for postgres, and I am using the wrong one?

I can't believe this, but I forgot to document the database changes from the previous Kuby release 🤦 Kuby used to support both Postgres and MySQL, but now only supports CockroachDB. There are a lot of good reasons for this, one of the biggest being that Cockroach is wire-compatible with Postgres. In other words they're both almost 100% compatible with each other. More importantly however, Cockroach has been designed to run in the cloud. You'll need to add the [cockroach-activerecord-adapter](https://github.com/cockroachdb/activerecord-cockroachdb-adapter) to your Gemfile and set the adapter to `cockroachdb` in your database.yml. I will update the docs right now.

**Comment by JorgenIvarsson** [comment]
Yes, that was strange and I have no explanation for not being able to connect to the app. I could l connect to the app through the command "bundle exec kuby -e production remote shell". I could even spin up the puma server with "Rails server". I did not configure a domain name.  Instead I tried to connect via the public ip-adress provided by the load balancer. So on one hand the app was running via the shell and I could also access the log files. And on the other hand I could not connet via ip-adress in the browser or curl. All I got was a 404 error message from nginx. Here is the docker image I used:
https://hub.docker.com/repository/docker/jorgenivarsson/kuby

So now I have a created a new fresh app that uses a database (Postgresql). If I want to use Kuby, do I have to switch to cockroach completely or can I use postgres as a development database and only use cockroach as a production database in my database.yml? Maybe load the gem file for cockroach under a production group? 
I'm not sure what happens when the app is dockerized? Is the database actually involved in the dockerization/kubernetization of the app or is only the database schema and migrations used? The reason I ask is that I already have mysql/mariadb and postgres installed on my computer. I rather not install (or learn) yet another database if I can avoid it.  I understand the reasons for choosing Cockroach. But the demands for using a specific database complicates things if there is no simple way of switching from say Postgres to Cockroach. Or is Cockroach a new default for all cloud based applications so it is expected that you use it together with Kubernetes from the start when developing a new application?
Anyway, thanks for your help and patience @camertron!

**Comment by camertron** [comment]
Hey @JorgenIvarsson, apologies for the delayed response.

>  If I want to use Kuby, do I have to switch to cockroach completely or can I use postgres as a development database and only use cockroach as a production database in my database.yml?

Yes, what you've described is absolutely possible. Kuby doesn't offer a solution for development environments at the moment.

> Maybe load the gem file for cockroach under a production group?

Yes, that's what I'd suggest 😄 

> Is the database actually involved in the dockerization/kubernetization of the app or is only the database schema and migrations used?

The schema and migrations are part of your app and are copied into the Docker image, and the CockroachDB adapter gem is installed inside the image as well. Other than that, the choice of database isn't considered during the Docker build.

Kuby instructs Kubernetes to launch several CockroachDB pods (older versions launched Postgres or MySQL pods). You can tell Kuby to not create any database pods (for example, if you're using a managed database offering from your cloud provider) by setting `manage_database false` as documented [here](https://getkuby.io/docs/customizing-deployments#the-rails-app-plugin).
